lib/front_matter.js

```diff
@@ -2,40 +2,36 @@
 
 const yaml = require('js-yaml');
 const rPrefixSep = /^(-{3,}|;{3,})/;
-const rFrontMatter = /^(-{3,}|;{3,})\n([\s\S]+?)\n\1(?:$|\n([\s\S]*)$)/;
-const rFrontMatterNew = /^([\s\S]+?)\n(-{3,}|;{3,})(?:$|\n([\s\S]*)$)/;
+const rFrontMatter = /^(-{3,}|;{3,})\n([\s\S]+?)\n\1\n?([\s\S]*)/;
+const rFrontMatterNew = /^([\s\S]+?)\n(-{3,}|;{3,})\n?([\s\S]*)/;
 
 function split(str) {
   if (typeof str !== 'string') throw new TypeError('str is required!');
 
-  if (rFrontMatter.test(str)) return splitOld(str);
-  if (rFrontMatterNew.test(str)) return splitNew(str);
-
-  return {content: str};
-}
+  const matchOld = str.match(rFrontMatter);
+  if (matchOld) {
+    return {
+      data: matchOld[2],
+      content: matchOld[3] || '',
+      separator: matchOld[1],
+      prefixSeparator: true
+    };
+  }
 
-function splitOld(str) {
-  const match = str.match(rFrontMatter);
+  if (rPrefixSep.test(str)) return { content: str };
 
-  return {
-    data: match[2],
-    content: match[3] || '',
-    separator: match[1],
-    prefixSeparator: true
-  };
-}
+  const matchNew = str.match(rFrontMatterNew);
```

Comment:
**Before**

- old: test old pattern -> match old pattern
- new: test old pattern -> match new pattern

**After**:

- old: match old pattern
- new: match old pattern -> match new pattern

As I found out in https://github.com/hexojs/hexo/pull/4436, there is no significant performance difference between `String#match` & `RegExp#test`.